### PR TITLE
HAWQ-577. Updated PXF metadata api to stream response

### PR DESCRIPTION
**Pull request by shivzone**

### CODE REVIEW

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 1]
```diff
@@ -46,76 +45,50 @@
...
+
+        if (null == metadataList || metadataList.isEmpty()) {
+            LOG.debug("No metadata");
+            return;
+        }
 
         for(Metadata metadata: metadataList) {
-            result.append(" Metadata for item \"");
+            StringBuilder result = new StringBuilder();
+            result.append("Metadata for item \"");
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: the quotes are not closed for the case of null metadata.
perhaps move `\"` to be part of line 79.

Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: good catch

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/rest/MetadataResource.java [hunk 2]
```diff
@@ -41,6 +41,7 @@
 import org.apache.hawq.pxf.api.MetadataFetcher;
 import org.apache.hawq.pxf.api.utilities.InputData;
 import org.apache.hawq.pxf.service.MetadataFetcherFactory;
+import org.apache.hawq.pxf.service.MetadataResponse;
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/rest/MetadataResource.java]: is this a new file?

Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/rest/MetadataResource.java]: yes, it is. I missed adding it and i've just updated

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        String prefix = "";
+        for (Metadata metadata : metadataList) {
+            if(metadata == null) {
+                throw new IllegalArgumentException("metadata object is null - cannot serialize");
+            }
+            if ((metadata.getFields() == null) || metadata.getFields().isEmpty()) {
+                throw new IllegalArgumentException("metadata for " + metadata.getItem() + " contains no fields - cannot serialize");
+            }
+            StringBuilder result = new StringBuilder();
+            /* metaData and userData are automatically converted to Base64 */
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: probably remove comment :)

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        DataOutputStream dos = new DataOutputStream(output);
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.setSerializationInclusion(Inclusion.NON_EMPTY); // ignore empty fields
+
+        if(metadataList == null || metadataList.isEmpty()) {
+            dos.write(METADATA_DEFAULT_RESPONSE.getBytes());
+            return;
+        }
+
+        dos.write("{\"PXFMetadata\":[".getBytes());
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: perhaps make PXFMetadata into a const, and use it here and as part of of METADATA_DEFAULT_RESPONSE.

**Comment by hornn** [comment]
+1. looks good

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java (hunk 4 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: I didn't want to do that. The default response can possibly change tomorrow and wanted to decouple that from a valid response string.

Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: You mean that the default might not contain the header "PXFMetadata"?
In that case I would do something like
```
final String METADATA_HEADER_RESPONSE = "PXFMetadata";
final String METADATA_DEFAULT_RESPONSE = "{\"" + METADATA_HEADER_RESPONSE + "\":[]}"
```

Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: What i meant was I wouldn't want to have any coorelation between metadata_default_response and the header string used in the actual response. I wouldn't mind introducing a separate constant for the string used in the acutal response but wouldn't use that as part of the default response

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        this.metadataList = metadataList;
+    }
+
+    /**
+     * Serializes the metadata list in JSON, To be used as the result string for HAWQ.
+     */
+    @Override
+    public void write(OutputStream output) throws IOException,
+            WebApplicationException {
+        DataOutputStream dos = new DataOutputStream(output);
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponse.java]: Should we close resource in the end?

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 6]
```diff
@@ -46,76 +45,49 @@
...
+            return;
+        }
 
         for(Metadata metadata: metadataList) {
-            result.append(" Metadata for item \"");
+            StringBuilder result = new StringBuilder();
 
             if (metadata == null) {
-                return "No metadata";
+                result.append("None");
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: Literal "None" was used couple times, might me a constant.

**Comment by sansanichfb** [comment]
LGTM

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java (hunk 6 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: This is just used for logging and not for the response. Defining constants just for log message seems an overkill

**Pull request closed without merging**